### ISSUE TITLE
Crashes on Windows on ARM

### ISSUE THREAD

**Issue by orditeck** [Martchus/syncthingtray#304]
**Bug description**
Every time I put my Surface Pro 11 to sleep and wake it up, the Syncthing Tray icon has disappeared and syncthing isn't running anymore. Doesn't happen with any other app.

Addition: it also just crashed after I right-clicked on the icon, so I guess it's not only after sleep.

Version: syncthingtray-1.6.4-aarch64-w64-mingw32
Qt 6.8.1

**Comment by orditeck** [comment]
Now using the x64 version, emulated for ARM, drop-in replacement and no crash so far.

**Comment by Martchus** [comment]
Sounds similar to what has been discussed on https://github.com/Martchus/syncthingtray/issues/218#issuecomment-2502550610. Probably it makes sense to have a distinct issue for that so I'm not closing this as duplicate.

Considering I don't have aarch64 hardware to reproduce this I'm limited in how I can help here. It makes probably sense if you debug this further in your own but I can of course try to assist.

Maybe it makes a difference that the ARM builds use libc++ instead of libstdc++.

**Comment by Martchus** [comment]
So I recommend you create a debug build using MSYS2's mingw-w64 packaging. You can find build instructions on the README of c++utilities (another repo on my GitHub page).

**Comment by Martchus** [comment]
I was able to compile the latest version of Boost for Windows on ARM with the Boost.Process library using the patch from https://github.com/msys2/MINGW-packages/commit/da4f4c37f534bfc47327a47a3eb9c0e9d6cd622a#diff-35ebcd8b8f3e30bb84ddbfad90b6dd1ed6f063ae4b2197b55464081b736f5a28R85. I was also able to compile the latest Syncthing Tray release against it using Boost.Process. The builds that are supposed to be statically linked are unfortunately dynamically linked so one can't test the binaries easily at the moment. (The linking problem has been fixed in c++utilities and thus will be fixed on the next release. Maybe I can provide development builds in the meantime. I also came up with a change in my release scripts to prevent uploading binaries that depend on DLLs not provided by Windows itself.)

**Comment by theAeon** [comment]
Currently running the arm build myself and have both VS devtools and mingw installed-I can try giving stuff a debug compile when i get the chance later.

**Comment by theAeon** [comment]
okay for whatever its worth it appears to have survived multiple sleep wake cycles w/ boost-process

will let you know if/when it dies

**Comment by theAeon** [comment]
yeah i think that was it

**Comment by Martchus** [comment]
So it works *with* Boost.Process? That's good news. I only hope that my cross-compiled builds using static libraries will work as well.

**Comment by theAeon** [comment]
That it does. Feel free to shoot over a test build it you'd like.

**Comment by Martchus** [comment]
https://martchus.dyn.f3l.de/repo/arch/ownstuff-experimental/os/x86_64/mingw-w64-clang-aarch64-syncthingtray-qt6-1.7.0-1-any.pkg.tar.zst

**Comment by theAeon** [comment]
....that one crashed on first wake.

**Comment by theAeon** [comment]
```
ModLoad: 00007ffc`86d50000 00007ffc`86d90000   C:\windows\SYSTEM32\gpapi.dll
(245ac.29144): C++ EH exception - code e06d7363 (first chance)
(245ac.29144): C++ EH exception - code e06d7363 (first chance)
(245ac.29144): C++ EH exception - code e06d7363 (first chance)
(245ac.29144): C++ EH exception - code e06d7363 (first chance)
ModLoad: 00007ffc`3c7e0000 00007ffc`3c974000   C:\windows\SYSTEM32\d3d9on12.dll
ModLoad: 00007ffc`3c020000 00007ffc`3c43e000   C:\windows\SYSTEM32\D3D12Core.dll
ModLoad: 00007ffc`3b400000 00007ffc`3bc77000   C:\windows\System32\DriverStore\FileRepository\qcdx8380.inf_arm64_cd283b9bc940b474\qcdx12arm64xum.dll
ModLoad: 00007ffb`c5ce0000 00007ffb`c5f40000   C:\windows\System32\DriverStore\FileRepository\qcdx8380.inf_arm64_cd283b9bc940b474\qcegparm64x.DLL
ModLoad: 00007ffc`3a6c0000 00007ffc`3a9a2000   C:\windows\SYSTEM32\dxilconv.dll
ModLoad: 00007ffc`67d20000 00007ffc`6a509000   C:\windows\System32\DriverStore\FileRepository\qcdx8380.inf_arm64_cd283b9bc940b474\qcgpuarm64xcompilercore.dll
ModLoad: 00007ffc`67a70000 00007ffc`67ac1000   C:\windows\SYSTEM32\D3DSCache.dll
ModLoad: 00007ffc`83cf0000 00007ffc`83d1a000   C:\windows\SYSTEM32\resourcepolicyclient.dll
ModLoad: 00007ffc`495c0000 00007ffc`4967b000   C:\windows\system32\dataexchange.dll
ModLoad: 00007ffc`72be0000 00007ffc`72e68000   C:\windows\SYSTEM32\textinputframework.dll
ModLoad: 00007ffc`7c5d0000 00007ffc`7c85b000   C:\windows\SYSTEM32\CoreMessaging.dll
ModLoad: 00007ffc`6c6c0000 00007ffc`6cd2f000   C:\windows\SYSTEM32\CoreUIComponents.dll
ModLoad: 00007ffc`49860000 00007ffc`4994c000   C:\windows\system32\Oleacc.dll
ModLoad: 00007ffc`49d50000 00007ffc`4a58d000   C:\windows\SYSTEM32\uiautomationcore.DLL
ModLoad: 00007ffc`88100000 00007ffc`88200000   C:\windows\SYSTEM32\sxs.dll
mincore\com\oleaut32\dispatch\ups.cpp(2126)\OLEAUT32.dll!00007FFC8A44DAB0: (caller: 00007FFC8A44F7EC) ReturnHr(1) tid(29144) 8002801D Library not registered.
(245ac.29144): C++ EH exception - code e06d7363 (first chance)
(245ac.29144): C++ EH exception - code e06d7363 (first chance)
(245ac.1dcc): Security check failure or stack buffer overrun - code c0000409 (!!! second chance !!!)
Subcode: 0xa FAST_FAIL_GUARD_ICALL_CHECK_FAILURE 
ntdll!RtlFailFast2:
00007ffc`8d02ee20 d43e0060 brk         #0xF003
```

exception analysis tbd

**Comment by Martchus** [comment]
Probably my builds are done with more security hardening flags than your builds so a problem with the code leads to this hard crash.

**Comment by theAeon** [comment]
want me to rebuild on my end w/ your flags?

**Comment by Martchus** [comment]
You could try; I'm not sure whether it'll make a difference when just compiling Syncthing Tray itself with different flags. Maybe the compilation of Qt, other dependencies and mingw-w64 itself is also relevant.

**Comment by Martchus** [comment]
It looks like the use of `--enable-cfguard` when compiling mingw-w64-crt is at least also present in MSYS2 (https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-crt-git/PKGBUILD#L51).

**Comment by theAeon** [comment]
```
    Run Build Command(s): C:/rtools44-aarch64/clangarm64/bin/ninja.exe -v cmTC_41189
    [1/2] C:\rtools44-aarch64\clangarm64\bin\cc.exe   -rtlib=compiler-rt -fuse-ld=lld -mguard=cf -D_FORTIFY_SOURCE=3 -D_GLIBCXX_ASSERTIONS -O2 -pipe -fno-plt -fexceptions --param=ssp-buffer-size=4 -Wformat -Werror=format-security -mguard=cf -fcf-protection -MD -MT CMakeFiles/cmTC_41189.dir/testCCompiler.c.obj -MF CMakeFiles\cmTC_41189.dir\testCCompiler.c.obj.d -o CMakeFiles/cmTC_41189.dir/testCCompiler.c.obj -c E:/build/CMakeFiles/CMakeScratch/TryCompile-tsfez2/testCCompiler.c
    FAILED: CMakeFiles/cmTC_41189.dir/testCCompiler.c.obj
    C:\rtools44-aarch64\clangarm64\bin\cc.exe   -rtlib=compiler-rt -fuse-ld=lld -mguard=cf -D_FORTIFY_SOURCE=3 -D_GLIBCXX_ASSERTIONS -O2 -pipe -fno-plt -fexceptions --param=ssp-buffer-size=4 -Wformat -Werror=format-security -mguard=cf -fcf-protection -MD -MT CMakeFiles/cmTC_41189.dir/testCCompiler.c.obj -MF CMakeFiles\cmTC_41189.dir\testCCompiler.c.obj.d -o CMakeFiles/cmTC_41189.dir/testCCompiler.c.obj -c E:/build/CMakeFiles/CMakeScratch/TryCompile-tsfez2/testCCompiler.c
    error: option 'cf-protection=return' cannot be specified on this target
    error: option 'cf-protection=branch' cannot be specified on this target

```

it won't even let me 😅

**Comment by R-Goc** [comment]
This stack trace doesn't even leave windows DLLs. Unless it is corrupted main wasn't even called.

**Comment by Martchus** [comment]
Maybe `PowerpResumeSuspendCallback` is calling some callback but the address of the callback is wrong. However, none of my code is using these WinAPIs trying to supply some kind of callback. Maybe some other piece of code in the software stack does (e.g. Qt, however, I doubt it and couldn't find anything after a brief search).

**Comment by R-Goc** [comment]
There is no user code in this stack trace.

**Comment by R-Goc** [comment]
Nevermind you're right. Rtlp­Handle­Invalid­User­Call­Target handles a call to an invalid function pointer.

**Comment by R-Goc** [comment]
If you could read the registers it would be nice. The function pointer should still be there.

**Comment by Martchus** [comment]
I suppose the relevant WinAPI function where one would pass the callback is PowerRegisterSuspendResumeNotification (https://learn.microsoft.com/en-us/windows/win32/api/powerbase/nf-powerbase-powerregistersuspendresumenotification). I'm just wondering where this function is called. I've already checked Qt, relevant Boost libraries and LLVM but none of these code bases contain `PowerRegisterSuspendResumeNotification`. Of course this function might also be called by any of the loaded DLLs for whatever reason.

**Comment by R-Goc** [comment]
That is my suspicion.

**Comment by theAeon** [comment]
on it.

**Comment by R-Goc** [comment]
Here is a blog from a similar failure. https://devblogs.microsoft.com/oldnewthing/20240913-00/?p=110257

**Comment by theAeon** [comment]
```
0:030> .frame /r
00 000000eb`c3aff0a0 00007ffc`8cfcb518     ntdll!RtlFailFast2
 x0=000000000000000a   x1=00007ff661298ec8   x2=0000000000000000   x3=000000ebc3aff4b0
 x4=000000ebc3aff4e0   x5=0000000000000000   x6=0000000000000000   x7=0000000000000000
 x8=0000000000000001   x9=00007ffc8d2584e0  x10=000002d6f610b788  x11=000002d6f610b788
x12=0000000000000000  x13=0000000000000000  x14=0000000000000000  x15=00007ff661298ec8
x16=0000000000000000  x17=00007ffc8d01eba0  x18=0000000000000000  x19=00007ff661298ec8
x20=0000000000000004  x21=0000000000000000  x22=000002d6f6131b00  x23=000000007ffe0008
x24=0000000000000000  x25=0000000000000000  x26=0000000000000001  x27=0000000000000000
x28=0000000000000000   fp=000000ebc3aff0a0   lr=00007ffc8cfcb518   sp=000000ebc3aff0a0
 pc=00007ffc8d02ee20  psr=80000040 N--- EL0
ntdll!RtlFailFast2:
00007ffc`8d02ee20 d43e0060 brk         #0xF003
```

**Comment by R-Goc** [comment]
No clue about arm calling convention. Not sure why that would be invalid. A few other look like function pointer as well.

**Comment by theAeon** [comment]
yeah i'm updating as i get them

**Comment by theAeon** [comment]
that's all of them- the rest were either duplicates or garbage

**Comment by R-Goc** [comment]
syncthingtray_qt6_static!pcre2_get_error_message_16+0xe41a8

This is the most suspicious one I guess. Not sure what regexes you are using. And why this would be invalid

**Comment by theAeon** [comment]
Or why pcre2 is at all involved in a power callback.

**Comment by Martchus** [comment]
Yes, this looks a little bit suspicious. PCRE2 is used by Qt to implement `QRegularExpression` which is only used in very few occations in Syncthing Tray and maybe in various places in Qt itself. I'm also wondering why it would be invalid and why it would be invoked via the callback. Maybe there's something is corruped even worse so `pcre2_get_error_message_16` is just a red herring.

(Syncthing Tray doesn't run any user-specified regexes so you can unfortunately not easly test whether processing invalid regexes will also trigger a crash.)

**Comment by R-Goc** [comment]
I guess brake inside of powrprof!PowerpResumeSuspendCallback and see what that ends up calling.

**Comment by theAeon** [comment]
putting aside that I have no idea why the regex is being invoked _at all_ i think that what's happening here is that the regex exception is triggering CFG. I'm curious what it's trying to call that would even do that.

**Comment by theAeon** [comment]
....there's no symbol.

```
powrprof!PowerpResumeSuspendCallback+0xb0:
00007ffc`883751e0 94008798 bl          powrprof!guard_check_icall$thunk$15021643654165956172 (00007ffc`88397040)
0:001> t
powrprof!guard_check_icall$thunk$15021643654165956172:
00007ffc`88397040 17ffab50 b           powrprof!guard_check_icall (00007ffc`88381d80)
0:001> t
powrprof!guard_check_icall:
00007ffc`88381d80 900000d0 adrp        xip0,powrprof!_os_arm64x_dispatch_call_no_redirect (00007ffc`88399000)
0:001> t
powrprof!guard_check_icall+0x4:
00007ffc`88381d84 f9403a10 ldr         xip0,[xip0,#0x70]
0:001> t
powrprof!guard_check_icall+0x8:
00007ffc`88381d88 d61f0200 br          xip0 {00007ffc`8d2e0040}
0:001> t
00007ffc`8d2e0040 d2800010 mov         xip0,#0
0:001> t
00007ffc`8d2e0044 f2b33fd0 movk        xip0,#0x99FE,lsl #0x10
0:001> t
00007ffc`8d2e0048 f2cfbeb0 movk        xip0,#0x7DF5,lsl #0x20
0:001> t
00007ffc`8d2e004c f2e00010 movk        xip0,#0,lsl #0x30
0:001> t
00007ffc`8d2e0050 d346fdf1 lsr         xip1,x15,#6
0:001> t
00007ffc`8d2e0054 f2400dff tst         x15,#0xF
0:001> t
00007ffc`8d2e0058 38716a11 ldrb        wip1,[xip0,xip1]
0:001> t
00007ffc`8d2e005c d34315f0 ubfx        xip0,x15,#3,#3
0:001> t
00007ffc`8d2e0060 54000081 bne         00007ffc`8d2e0070
0:001> t
00007ffc`8d2e0070 927ffa10 and         xip0,xip0,#-2
0:001> t
00007ffc`8d2e0074 9ad02631 lsr         xip1,xip1,xip0
0:001> t
00007ffc`8d2e0078 36000051 tbz         wip1,#0,00007ffc`8d2e0080
0:001> t
00007ffc`8d2e007c 370fff91 tbnz        wip1,#1,00007ffc`8d2e006c
0:001> t
00007ffc`8d2e0080 d2800010 mov         xip0,#0
0:001> t
00007ffc`8d2e0084 1400006f b           00007ffc`8d2e0240
0:001> t
00007ffc`8d2e0240 d29d7411 mov         xip1,#0xEBA0
0:001> t
00007ffc`8d2e0244 f2b1a031 movk        xip1,#0x8D01,lsl #0x10
0:001> t
00007ffc`8d2e0248 f2cfff91 movk        xip1,#0x7FFC,lsl #0x20
0:001> t
00007ffc`8d2e024c f2e00011 movk        xip1,#0,lsl #0x30
0:001> t
00007ffc`8d2e0250 d61f0220 br          xip1 {ntdll!LdrpHandleInvalidUserCallTarget (00007ffc`8d01eba0)}
0:001> t
ntdll!LdrpHandleInvalidUserCallTarget:
00007ffc`8d01eba0 a9b17bfd stp         fp,lr,[sp,#-0xF0]!
```

**Comment by theAeon** [comment]
```
0:036> u @x15 L1
syncthingtray_qt6_static!pcre2_get_error_message_16+0xe41a8:
00007ff6`61298ec8 b24003ec movi        x12,#1
```

just kidding!

**Comment by theAeon** [comment]
I do have to wonder if the reason that clangarm64 won't let me use the fcf-protection flag for a native build is because this is known to not work.

**Comment by R-Goc** [comment]
Looks probable. Here is a debian discussion for it. https://lists.debian.org/debian-devel/2022/10/msg00314.html. Also GCC docs document fcf-protection as using Intel CET.

**Comment by Martchus** [comment]
At least the flag when compiling mingw-w64 shouldn't hurt as MSYS2 enabled it as well. They also have tests using `-mguard=cf` which probably run on all archs. See https://github.com/msys2/MINGW-packages/issues/16833 for details.

It also looks like llvm-mingw enables it for all archs: https://github.com/mstorsjo/llvm-mingw/blob/master/Dockerfile

**Comment by theAeon** [comment]
probably worth noting that mguard=cf, enable-cfguard, and fcf-protection are all different flags

**Comment by theAeon** [comment]
only one that clangarm is giving me trouble for using is fcf

**Comment by Martchus** [comment]
I know but I'm also not using fcf. Sorry for the confusion, I initially thought as well that fcf might be the culprit but it is not used at all in my builds as it is only used when the condition `[[ $_arch_ == aarch64 ]]` is *not* true (see https://github.com/Martchus/PKGBUILDs/blob/9ab7f046f27d9abfade7efdb5c31c9b39d16746d/environment/mingw-w64-clang/mingw-env.sh#L13).

**Comment by theAeon** [comment]
Bash syntax, love it.

gonna see what happens if its a personal build w/o fcf then

**Comment by theAeon** [comment]
misaligned ldr/str offset?

**Comment by theAeon** [comment]
```
C:\windows\system32\cmd.exe /C "cd . && C:\rtools44-aarch64\clangarm64\bin\c++.exe -rtlib=compiler-rt -fuse-ld=lld -mguard=cf -D_FORTIFY_SOURCE=3 -D_GLIBCXX_ASSERTIONS -O2 -pipe -fno-plt -fexceptions --param=ssp-buffer-size=4 -Wformat -Werror=format-security -mguard=cf -stdlib=libc++ -g -rtlib=compiler-rt -fuse-ld=lld -mguard=cf -Wl,-O1,--sort-common,--as-needed -fstack-protector syncthingtray/cli/CMakeFiles/syncthingctl-cli.dir/syncthingctl-cli_autogen/mocs_compilation.cpp.obj syncthingtray/cli/CMakeFiles/syncthingctl-cli.dir/resources/windows-cli-wrapper-Debug.rc.obj syncthingtray/cli/CMakeFiles/syncthingctl-cli.dir/E_/syncthing/c++utilities/cmake/templates/cli-wrapper.cpp.obj -o syncthingtray\cli\syncthingctl-cli.exe -Wl,--out-implib,syncthingtray\cli\libsyncthingctl-cli.dll.a -Wl,--major-image-version,0,--minor-image-version,0 -LE:/build/c++utilities   -LE:/build/qtutilities   -LE:/build/syncthingtray/syncthingconnector   -LE:/lib -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && cd ."
ld.lld: error: misaligned ldr/str offset
ld.lld: error: misaligned ldr/str offset
ld.lld: error: misaligned ldr/str offset
ld.lld: error: misaligned ldr/str offset
ld.lld: error: misaligned ldr/str offset
c++: error: linker command failed with exit code 1 (use -v to see invocation)
[2/74] Linking CXX executable syncthingtray\tray\syncthingtray-cli.exe
FAILED: syncthingtray/tray/syncthingtray-cli.exe
C:\windows\system32\cmd.exe /C "cd . && C:\rtools44-aarch64\clangarm64\bin\c++.exe -rtlib=compiler-rt -fuse-ld=lld -mguard=cf -D_FORTIFY_SOURCE=3 -D_GLIBCXX_ASSERTIONS -O2 -pipe -fno-plt -fexceptions --param=ssp-buffer-size=4 -Wformat -Werror=format-security -mguard=cf -stdlib=libc++ -g -rtlib=compiler-rt -fuse-ld=lld -mguard=cf -Wl,-O1,--sort-common,--as-needed -fstack-protector syncthingtray/tray/CMakeFiles/syncthingtray-cli.dir/syncthingtray-cli_autogen/mocs_compilation.cpp.obj syncthingtray/tray/CMakeFiles/syncthingtray-cli.dir/resources/windows-cli-wrapper-Debug.rc.obj syncthingtray/tray/CMakeFiles/syncthingtray-cli.dir/E_/syncthing/c++utilities/cmake/templates/cli-wrapper.cpp.obj -o syncthingtray\tray\syncthingtray-cli.exe -Wl,--out-implib,syncthingtray\tray\libsyncthingtray-cli.dll.a -Wl,--major-image-version,0,--minor-image-version,0 -LE:/build/c++utilities   -LE:/build/qtutilities   -LE:/build/syncthingtray/syncthingconnector   -LE:/build/syncthingtray/syncthingmodel   -LE:/lib   -LE:/build/lib -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && cd ."
ld.lld: error: misaligned ldr/str offset
ld.lld: error: misaligned ldr/str offset
ld.lld: error: misaligned ldr/str offset
ld.lld: error: misaligned ldr/str offset
ld.lld: error: misaligned ldr/str offset
c++: error: linker command failed with exit code 1 (use -v to see invocation)
[3/74] Linking CXX static library syncthingtray\syncthingconnector\libsyncthingconnector.a
ninja: build stopped: subcommand failed.
```

**Comment by R-Goc** [comment]
This looks curious: 
https://github.com/llvm/llvm-project/issues/110186

**Comment by theAeon** [comment]
guess the binutils as compiled by the msys guys aren't above board here, huh

**Comment by Martchus** [comment]
The issue with `ld.lld: error: misaligned ldr/str offset` was very strange, indeed. I came to the conclusion this was just due bintuils and llvm tooling being incompatible. That you're now running into the issue as well is even more curious and maybe my previous conclusion was wrong.

Note that I rebuilt all my packages using llvm tooling. (The only package where I used binutils was boot but only for windmc. Other tooling was taken from llvm. I think it would have been impossible to use binutils anyways because the misalignment issue goes in the way very soon.)

**Comment by theAeon** [comment]
perhaps even more strange is that this wasn't happening until i started adding the hardening flags.

**Comment by theAeon** [comment]
culprit appears to either be fno-plt or fexceptions, per testing

**Comment by theAeon** [comment]
```
0:035> u @x1 L1
syncthingtray!runtime.callbackasm+0x8 [./runtime/zcallback_windows_arm64.s @ 15]:
00007ff6`d7d29b88 b24003ec movi        x12,#1
```
that's....new.

**Comment by Martchus** [comment]
Ok, so what flags exactly did you have to add to the compilation of what to reproduce a crash? And did you reproduce the power callback crash or only the one pointing to the Go runtime? Unfortunately I'm really not familiar with what these callback from the Go runtime to be able to shed some light on what's perhaps going on. Maybe you're onto something, though :-)

**Comment by theAeon** [comment]
> Ok, so what flags exactly did you have to add to the compilation of what to reproduce a crash? And did you reproduce the power callback crash or only the one pointing to the Go runtime? Unfortunately I'm really not familiar with what these callback from the Go runtime to be able to shed some light on what's perhaps going on. Maybe you're onto something, though :-)

Same flags minus -fno-plt. That's the function pointer present when the same powercallback trips.

**Comment by Martchus** [comment]
And I assume you're only recompiling Syncthing Tray but none of the dependencies like Qt, libc++ or mingw-w64? If the compilation of Syncthing Tray alone makes a difference that's maybe a good sign (and I won't have to recompile everything to create a fixed build).

To exclude Go you can of course build with `NO_LIBSYNCTHING=ON`. and `USE_LIBSYNCTHING=OFF`.

**Comment by theAeon** [comment]
Correct. Not sure how to get the output of a no_libsyncthing build to attach to a running copy of syncthing, though.

**Comment by theAeon** [comment]
seems its surviving sleep-wake with external syncthing. next testing your current release build.

**Comment by Martchus** [comment]
Juding by https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html#index-fno-plt and other information I found about `-fno-plt` I am starting to think that this flag is not relevant for PE targets as it seems rather ELF specific. Unfortunately I haven't found a clear statement that it is ELF specific yet. However, if that's true then it would be very strange if it was the culprit.

I think I only added this flag because it is also used in regular x86 mingw-w64 builds (https://aur.archlinux.org/cgit/aur.git/tree/mingw-env.sh?h=mingw-w64-environment). It might just ended up being used in regular x86 mingw-w64 builds because we generally try to stay close to how things are done in native Arch Linux packaging (to have the same level of hardening and optimization).

**Comment by theAeon** [comment]
still dies. seems its a function of whether libsyncthing is compiled in rather than a function of whether its actually using it.

(also interesting that this ones back to the libpcre error)

**Comment by theAeon** [comment]
> However, if that's true then it would be very strange if it was the culprit.
To be fair this didn't affect the crashing (although it did change the called-back function to something that at least made some sense). It only affected the alignment error on native build.

**Comment by Martchus** [comment]
> still dies. seems its a function of whether libsyncthing is compiled in rather than a function of whether its actually using it.

That doesn't surprise me. A few years ago I had crashes when trying to run an x86 Windows build with Wine. It always crashed when it was compiled with libsyncthing support, even though none of the Go functions were ever executed. (It worked natively under Windows, though.)

**Comment by Martchus** [comment]
I asked about this on the LLVM discord and got answers:

* `-fno-plt` has no effect on Windows/PE targets, indeed.
* `-mguard=cf` is supported by LLVM for mingw-w64 targets including aarch64 as long as mingw-w64 is built with `--enable-cfguard` (my build of mingw-w64 is configured with that flag).

About debugging the linker errors:

> Oh btw, if you or the reporter run into issues with LLD, regarding the issues with alignment - rerun the link command and add `-Wl,--reproduce=repro.tar`, this produces a tarball which can be shared, which should allow someone else (me) to reproduce and debug what's really going on at the link stage, without needing to share the whole compilation/build environment

**Comment by Martchus** [comment]
I've just released 1.7.0 and for this release I removed support for the built-in Syncthing library in ARM builds. Maybe this makes a difference.

**Comment by theAeon** [comment]
Appears it survives sleep-wake.

Dunno if that counts as solving the issue, but it certainly solves my usecase.

**Comment by Martchus** [comment]
I had one new thought: Maybe I just need to be more careful with what flags to pass to cgo specifically. My CMake-based build system passes flags via the the env variables `CGO_CFLAGS`, `CGO_CXXFLAGS` and `CGO_LDFLAGS` from CMake to cgo. The idea is to build everything with consistent flags that are specified in one place (CMake level e.g. from toolchain file). I also had a brief look at the code and I must say that passing the flags is probably actually slightly broken. So there's definitely room for improvement anyway.

**Comment by Martchus** [comment]
Note that issues with `ld.lld: error: misaligned ldr/str offset` were definitely caused by use of `-fno-plt`, see https://github.com/Martchus/PKGBUILDs/commit/dcbd1692139ed131468220e24de3e99c7c6d366b for an explanation and further links.

I plan to do the next release once Qt 6.8.2 is out (should have been released yesterday). Then the builds for Windows on ARM will be using LLVM/Clang/libc++ 19.1.7. For this release I will keep support for the built-in Syncthing library disabled. However, I will also build additional binaries with the build-in Syncthing library enabled:

1. A version where the mguard flag is not passed to the Go build. Maybe this already helps to prevent the issue.
2. A version where the mguard flag is not used at all during the build of Syncthing Tray. If 1. does not prevent the issue then maybe not using mguard at all helps. However, my build of mingw-w64 and all other dependencies will still  use mguard flag. So maybe this doesn't make a big difference.
3. A version where the mguard flag is still used everywhere as before. Maybe it will just work after updating the LLVM and Go toolchains.

**Comment by Martchus** [comment]
By the way, I recently dealt with [another problem caused by a limitation of the Go runtime](https://github.com/Martchus/syncthingtray/commit/a81e6d47521098ec6473c345109b5cb63a381621) affecting Go program built as a library on certain platforms. So maybe we are hitting another limitation here as well that is possibly specific to the program being built as a library and the use of mguard.

**Comment by Martchus** [comment]
I've just released a new version of Syncthing Tray. As mentioned in https://github.com/Martchus/syncthingtray/issues/304#issuecomment-2613524299 I compiled a few additional binaries for Windows on ARM for experimentation. It would be great if you could test them.

A version without Go / built-in Syncthing can be found on the website. That is hopefully working.

It would also be great if one could test the other versions with Go / built-in Syncthing from https://martchus.dyn.f3l.de/repo/arch/ownstuff-experimental/os/x86_64 and report how the behave on suspend/resume. Run the executables with `-static` suffix (as the other ones require additional DLLs).

**Comment by Martchus** [comment]
Can you try these additional binaries as mentioned in the previous comments and give feedback?

**Comment by theAeon** [comment]
yeah-sorry, been busy

**Comment by alvinhochun** [comment]
Hi, I stumbled across this issue since one of the comments mentioned another issue in MSYS. I thought I should give some of my thoughts:

> ```
> 0:036> u @x15 L1
> syncthingtray_qt6_static!pcre2_get_error_message_16+0xe41a8:
> 00007ff6`61298ec8 b24003ec movi        x12,#1
> ```
Indeed, CFGuard on ARM64 passes the address of an indirect call to `guard_check_icall` by `x15`, according to LLVM.

---

I would say this likely has *nothing* to do with pcre2 or regex. Even though it references the symbol `pcre2_get_error_message_16`, note that the offset is `0xe41a8`, which is huge. You don't usually get functions this huge, so I would guess that the address is way beyond the end of the `pcre2_get_error_message_16` function.

I am guessing `syncthingtray_qt6_static` is a statically linked EXE? If it has DWARF debugging symbols you can use addr2line (or llvm-addr2line) to find its code location.

(As to why WinDbg shows this symbol name, my hypothesis is that libpcre2 incorrectly set `__declspec(dllexport)` on some functions even when building a static build, which causes them to end up in the export table of the EXE. WinDbg was simply finding the closest export symbol and showing the offset to it.)

**Comment by Martchus** [comment]
@alvinhochun Thanks for sharing your thoughts. They are in-line with what I suspect at this point as well.

> I would say this likely has *nothing* to do with pcre2 or regex. Even though it references the symbol `pcre2_get_error_message_16`, note that the offset is `0xe41a8`, which is huge. You don't usually get functions this huge, so I would guess that the address is way beyond the end of the `pcre2_get_error_message_16` function.

Yes, this function coming up is probably a red herring.

> I am guessing `syncthingtray_qt6_static` is a statically linked EXE? If it has DWARF debugging symbols you can use addr2line (or llvm-addr2line) to find its code location.

Yes, it links against all libraries not provided by Windows itself statically. It has no debugging symbols, though.

> (As to why WinDbg shows this symbol name, my hypothesis is that libpcre2 incorrectly set `__declspec(dllexport)` on some functions even when building a static build, which causes them to end up in the export table of the EXE. WinDbg was simply finding the closest export symbol and showing the offset to it.)

That makes sense. I am also aware that some libraries in the dependency stack wrongly set `__declspec(dllexport)` on static builds leading to unnecessary exports in the final exe.

**Comment by theAeon** [comment]
> Can you try these additional binaries as mentioned in the previous comments and give feedback?

Funnily enough, the mingw-w64-clang-aarch64-syncthingtray-qt6-libst-1.7.2-1-any build seems to be surviving sleep just fine.

edit: and she's dead

**Comment by theAeon** [comment]
syncthingtray-qt6-libst-no-mguard-cgo-1.7.2-1-any does not survive sleep.

mingw-w64-clang-aarch64-syncthingtray-qt6-libst-no-mguard-all-1.7.2-1-any.pkg.tar.zst is behind http authentication for some reason

**Comment by Martchus** [comment]
> mingw-w64-clang-aarch64-syncthingtray-qt6-libst-no-mguard-all-1.7.2-1-any.pkg.tar.zst is behind http authentication for some reason

This should be fixed now. (A regex was not specific enough.)

Ok, so sooner or later all builds crashed except for the `*-all` version which you haven't tested yet. Unfortunately that's not very promising so far.

**Comment by theAeon** [comment]
That one died too.

**Comment by Martchus** [comment]
So if disabling mguard helps then probably only if I would recompile everything without it. I guess for now I'll just keep libsyncthing disabled.

**Comment by lapo-luchini** [comment]
Still happening, or happening again. (?)

> Now using the x64 version, emulated for ARM, drop-in replacement and no crash so far.

Work-around works for me too:
```
scoop uninstall syncthingtray
scoop install -a 64bit syncthingtray
```

**Comment by Martchus** [comment]
We previously established that the crashes are related to the builtin Syncthing library. Hence I disabled this feature for aarch64 and it is still enabled now.

Maybe you found a different type of crash. Maybe this different type of crash also happens only randomly and isn't aarch64-specific. Without further information this is hard to tell. Without more information about the crash there's is actually nothing I can do about it at all.

**Comment by Martchus** [comment]
I've just noticed a bug in the OpenSSL-based signature verification of the updater. This affects the ARM builds for Windows as I had to disable libsyncthing/Go there and it is therefore using OpenSSL instead. I have already fixed it but the newly added updater of the current release will show a signature verification error when trying to update. (The problem is quite stupid. OpenSSL apparently only accepts pem blocks with `BEGIN PUBLIC KEY` but not `BEGIN EC PUBLIC KEY` in contrast to libsyncthing/Go - but I accidentally used the latter.)

**Comment by theAeon** [comment]
Oh, youre gonna love this one. 2.0 spins forever w/o starting. Can't quite tell why.

**Comment by Martchus** [comment]
What exactly is spinning forever? Syncthing Tray or Syncthing itself? Any logs to observe? On x86_64 the Windows builds and built-in Syncthing library worked just fine.

**Comment by theAeon** [comment]
SyncthingTray. Zero feedback. Downgraded to 1.7.10 and its still happening so something else is up. Will report back.

edit: 1.7.7 works still, will try intermediates

edit 2: 1.7.8 works (of course it does, its 99.9% the same)

Something between 1.7.8 and 1.7.10 broke it.

**Comment by Martchus** [comment]
Strange. Not sure right now what change it could be.

So you're starting Syncthing Tray and nothing happens at all - no tray icon, nothing - the process is just stuck? Is it utilizing the CPU while being stuck?

**Comment by theAeon** [comment]
> Strange. Not sure right now what change it could be.
> 
> So you're starting Syncthing Tray and nothing happens at all - no tray icon, nothing - the process is just stuck? Is it utilizing the CPU while being stuck?

Correct, nothing. And it does seem to be using CPU.

**Comment by xgdgsc** [comment]
Yes. Also stuck on 1.7.7

**Comment by xgdgsc** [comment]
Same here.

**Comment by Martchus** [comment]
Maybe I did a problematic update of LLVM, mingw-w64 or some other dependency between 1.7.8 and 1.7.10. (My server with the old builds it currently down for maintenance which is also why I don't have time for investigating this.)

**Comment by Martchus** [comment]
I did some tests with WINE/qemu. It seems that the problem of the app being stuck at startup is reproducible as of 1.7.9.

I checked the BUILDINFO diff and the most notable update is LLVM and libc++ which was updated from 19.1.7 to 20.1.7. However, I also updated mingw-w64 from 12 to 13 and Qt 6 from 6.9.0 to 6.9.1. Considering `--help` still works in my tests with WINE/qemu it might be related to Qt.

My password manager is also affected but also only the GUI. Its CLI doesn't use Qt at all but contains non-trivial code (compared to `--help`) which uses exceptions, libc++ and OpenSSL and several Windows APIs directly. And all of this still works. So this speaks for Qt as the culprit.

**Comment by Martchus** [comment]
For completeness, here's the diff I mentioned: [BUILDINFO-1.7.8-to-1.7.9.txt](https://github.com/user-attachments/files/21992330/BUILDINFO-1.7.8-to-1.7.9.txt)

I'm currently in progress of updating Qt to 6.9.2. Maybe that will help with this, too.

**Comment by JWWolstenholme** [comment]
> SyncthingTray. Zero feedback. Downgraded to 1.7.10 and its still happening so something else is up. Will report back.
> 
> edit: 1.7.7 works still, will try intermediates
> 
> edit 2: 1.7.8 works (of course it does, its 99.9% the same)
> 
> Something between 1.7.8 and 1.7.10 broke it.

Experiencing these exact symptoms on a surface pro 11 for any version after `1.7.8` too (`aarch64-w64-mingw32`). I can't speak for the resuming from sleep crash issue. Windows version `24H2 26100.4946`. There're no logs that I can find to share. I'll keep an eye on this thread. Happy to help test stuff.

**Comment by Martchus** [comment]
I added some debug printing around the `QApplication` initialization:

```
    // init application
    std::cerr << "before qapp\n";
    auto application = QApplication(argc, argv);
    std::cerr << "after qapp\n";
```

Only the "before …" message is printed. So it gets stuck in Qt code making it more likely that a change in Qt caused this. The problem stays if I just use `QCoreApplication`. This is with Qt 6.9.2 now so the update didn't help.

**Comment by theAeon** [comment]
guess there's a couple of variations that we can try here-compile against older msys qt and/or compile against native msvc qt

**Comment by Martchus** [comment]
Yes. However, it is likely that MSYS2 and MSVC builds both work and just my static cross-compiled build has this problem. So I probably have to find out what's wrong with my build. I have already checked https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-qt6-base/PKGBUILD for any critical aarch64-specific patches I might be missing but couldn't find anything.

**Comment by Martchus** [comment]
I changed the code to spawn a console and used the `-cli.exe` to display it on the terminal (still all with WINE/qemu). This gets me this error message:

```
qt.qpa.screen: "Unable to open monitor interface to \\\\.\\DISPLAY1:" "Invalid window handle."
```

But it might be a wrong lead considering it already gets stuck with `QCoreApplication` and then this message isn't printed. Otherwise there are no log messages at all.

I also let it print the platform name (`QGuiApplication::platformName()`) and it prints `windows` as expected (and *not* e.g. offscreen which would explain the behavior).

**Comment by theAeon** [comment]
attempted a msvc build vs. qt 6.10b3, got one of windows's many esoteric library input errors, realized its trying to import QtOpenGL which is...not distributed with the official builds for reasons i don't quite understand.

**Comment by Martchus** [comment]
> attempted a msvc build vs. qt 6.10b3, got one of windows's many esoteric library input errors, realized its trying to import QtOpenGL which is...not distributed with the official builds for reasons i don't quite understand.

Strange; I cannot really help with msvc, though.

With some debug printing I was able to determine that it hangs within the function `QWindowsMessageWindowClassContext::QWindowsMessageWindowClassContext()` of Qt Core. When trying to figure out the exact line the additional debug printing unfortunately changed the behavior so the overall `QCoreApplication` constructor doesn't hang anymore. So this is probably a timing thing. With the debug printing it then hangs later on the first `setStyleSheet()` call in application code. And that function hangs within `QWidgetPrivate::setStyle_helper()` which hangs in turn on `QCoreApplication::sendEvent(q, &e);`. So it all comes back to hanging on the `QCoreApplication` which then internally seems to hang on `qApp->notify(receiver, event);`.

**Comment by theAeon** [comment]
Yeah, I know how to fix it, its just a pain.

**Comment by theAeon** [comment]
Update from MSVC syncthingtray built against Qt6.10.0b3 with recompiled qtbase and qtdeclarative w/ re-enabled opengl....works great.

this is from sha af6e12bf217e8c5e32d114edd38a27704cf069b3, not latest. Forgot to re-pull, but its recent enough.

edit: repulled, works.

**Comment by theAeon** [comment]
Let me see if I can't give it a static rebuild and upload, hold on.

might as well test it with built in syncthing too

edit: looks like you cant static link the msvc build? sad.

**Comment by theAeon** [comment]
https://robbinsa.me/files/syncthingtray-2.0.1-qt-6.10b3-arm64-win-msvc-nost.zip

for your testing pleasure. Will upload the one with bundled st when i build it.

**Comment by Martchus** [comment]
> Update from MSVC syncthingtray built against Qt6.10.0b3 with recompiled qtbase and qtdeclarative w/ re-enabled opengl....works great.

Great and also expected. Considering my previous findings the problems with my build are entirely related to my LLVM/Clang/mingw-w64 setup and not to the code. (All the functions mentioned in https://github.com/Martchus/syncthingtray/issues/304#issuecomment-3250926829 also haven't changed between the working and the first not working build of Qt so it also isn't Qt.)

> edit: looks like you cant static link the msvc build?

I'm sure you can but you would need to make a static build of Qt yourself (as I think upstream Qt only provides shared objects).

> https://robbinsa.me/files/syncthingtray-2.0.1-qt-6.10b3-arm64-win-msvc-nost.zip
> for your testing pleasure.

Maybe someone who actually has an aarch64 system can use that while I'm still figuring out what's wrong with my build env. It would be great if your build was at least at 4c368af82e5d5d11be2dbd693e4cd54dd3f2add4, though - because that fixes a rather critical issue in the launcher.

> Will upload the one with bundled st when i build it.

And if you managed to do that, tell me how you did it :-)

I was trying to build Syncthing itself with MSVC (of course with the Go compiler but using MSVC for the native C bits) but it didn't work. Hence I changed the build script so that it would always use the default C compiler that Go would choose (which is GCC in the MSYS2 environment I installed Go in) despite using MSVC for the overall build. This also didn't work, though - the build ran into linker errors in the end. So you might need/want to revert https://github.com/Martchus/syncthingtray/blob/bf51eb32b749ca1b6944065f55ad3ee2fb2cbda4/syncthing/CMakeLists.txt#L116 and give using MSVC for everything another try. Another approach would be to build the Syncthing library as shared library. I meanwhile introduced code for this because under Android this is the only way. So maybe an extra condition like https://github.com/Martchus/syncthingtray/blob/bf51eb32b749ca1b6944065f55ad3ee2fb2cbda4/syncthing/CMakeLists.txt#L223 but for MSVC would make sense. With Syncthing v2 you might also want `-DSYNCTHING_USE_SYSTEM_SQLITE=OFF` unless you want to supply SQLite3 libraries externally.

**Comment by Martchus** [comment]
By the way, I rebuilt winpthreads and compiler-rt and libc++ and Qt and the problem was still reproducible in the exact same way as before. But I also don't think it is winpthreads because I only added that library for abseil-cpp and it isn't really used by Qt. (I removed it manually from the final linker line of the static executable and it still linked without errors.) However, I could be wrong. So if someone has further insights on this I'd be curious to learn more about the use of winpthreads in a mingw-w64/LLVM/Clang/libc++/aarch64 env.

**Comment by theAeon** [comment]
>Maybe someone who actually has an aarch64 system can use that while I'm still figuring out what's wrong with my build env. It would be great if your build was at least at https://github.com/Martchus/syncthingtray/commit/4c368af82e5d5d11be2dbd693e4cd54dd3f2add4, though - because that fixes a rather critical issue in the launcher.

I mean, I do! Mostly building it for my own use before anything else.

I did pull before building again, so that commit is in.

>I was trying to build Syncthing itself with MSVC (of course with the Go compiler but using MSVC for the native C bits) but it didn't work. Hence I changed the build script so that it would always use the default C compiler that Go would choose (which is GCC in the MSYS2 environment I installed Go in) despite using MSVC for the overall build. This also didn't work, though - the build ran into linker errors in the end. So you might need/want to revert 

yeah, it seems CGO _needs_ a gcc style command line so i've been using clang w/ msvc backend. It also appears it needs ar.exe to be present regardless of the variable AR.

bizarrely enough, they seem to be building but it never gets to linking syncthing into tray

**Comment by Martchus** [comment]
> bizarrely enough, they seem to be building but it never gets to linking syncthing into tray

You probably just miss `-DUSE_LIBSYNCTHING=ON`.

> so i've been using clang w/ msvc backend

That seems like a good idea. I'd be curious about the exact CMake arguments you used. It would make sense to add them to some of my README files. If you needed any changes in the CMakeLists.txt file we can probably upstream them, too.

**Comment by theAeon** [comment]
> You probably just miss -DUSE_LIBSYNCTHING=ON.

looks it. oops. will report back.

> That seems like a good idea. I'd be curious about the exact CMake arguments you used. It would make sense to add them to some of my README files. If you needed any changes in the CMakeLists.txt file we can probably upstream them, too.

No changes needed, just set env CC to clang and CXX to clang++ and copy llvm-ar to ar

**Comment by theAeon** [comment]
```
 Automatic MOC and UIC for target syncthingtray
     Creating library E:/syncthingtray/build/tray/Release/syncthingtray.lib and object E:/syncthingtray/build/tray/Rele
  ase/syncthingtray.exp
syncthing.lib(interface.obj) : error LNK2019: unresolved external symbol "void (__cdecl* libst_logging_callback_functio
n)(int,char const *,unsigned __int64)" (?libst_logging_callback_function@@3P6AXHPEBD_K@ZEA) referenced in function "pub
lic: __cdecl LibSyncthing::RunningState::RunningState(void)" (??0RunningState@LibSyncthing@@QEAA@XZ) [E:\syncthingtray\
build\tray\syncthingtray.vcxproj]
    Hint on symbols that are defined and could potentially match:
      libst_logging_callback_function
libsyncthinginternal.a(000015.o) : error LNK2019: unresolved external symbol _sqlite3_prepare_v2_blocking referenced in
 function _cgo_662e5d902139_Cfunc__sqlite3_prepare_v2_internal [E:\syncthingtray\build\tray\syncthingtray.vcxproj]
libsyncthinginternal.a(000015.o) : error LNK2019: unresolved external symbol _sqlite3_step_blocking referenced in funct
ion _cgo_662e5d902139_Cfunc__sqlite3_step_internal [E:\syncthingtray\build\tray\syncthingtray.vcxproj]
libsyncthinginternal.a(000015.o) : error LNK2019: unresolved external symbol _sqlite3_step_row_blocking referenced in f
unction _cgo_662e5d902139_Cfunc__sqlite3_step_row_internal [E:\syncthingtray\build\tray\syncthingtray.vcxproj]
E:\syncthingtray\build\tray\Release\syncthingtray.exe : fatal error LNK1120: 4 unresolved externals [E:\syncthingtray\b
uild\tray\syncthingtray.vcxproj]
```

So the 3 sqlite symbols look like they literally aren't exported by my vcpkg-built sqlite link, which is odd but a relatively easy fix, probably.

The one libst export looks like a missed `extern c` somewhere but I could be wrong.

**Comment by theAeon** [comment]
>The one libst export looks like a missed extern c somewhere but I could be wrong.

yup

```
diff --git a/syncthing/interface.cpp b/syncthing/interface.cpp
index d947d333..a27f4b78 100644
--- a/syncthing/interface.cpp
+++ b/syncthing/interface.cpp
@@ -15,8 +15,9 @@
 #define _CRT_USE_C_COMPLEX_H
 #endif

+extern "C" {
 #include "libsyncthinginternal.h"
-
+}
 #include <atomic>
 #include <cstring>
```

**Comment by JWWolstenholme** [comment]
> for your testing pleasure. Will upload the one with bundled st when i build it.

Works for me on a surface pro 11👍

**Comment by Martchus** [comment]
Strange, `libsyncthinginternal.h` contains

```
#ifdef __cplusplus
extern "C" {
#endif
```

before declaring any functions. At least in my build envs.

Not sure whether one can nest `extern "C" { }` blocks. If it is possible your change could nevertheless be upstreamed. Or is MSVC perhaps not providing the `__cplusplus` macro?

**Comment by theAeon** [comment]
> Strange, `libsyncthinginternal.h` contains
> 
> ```
> #ifdef __cplusplus
> extern "C" {
> #endif
> ```
> 
> before declaring any functions. At least in my build envs.
> 
> Not sure whether one can nest `extern "C" { }` blocks. If it is possible your change could nevertheless be upstreamed. Or is MSVC perhaps not providing the `__cplusplus` macro?

My libsyncthinginternal is missing that ifdef. I don't think you can nest them, but you can undefine __cplusplus, put your own extern C there, and redefine __cplusplus to handle both cases.

**Comment by theAeon** [comment]
Correction: mine contains that ifdef. I guess you can nest them because by calling something with extern "C" it auto-undefines it.

What I don't quite know is why it requires the wrapping outside, but it does fix the problem.

**Comment by theAeon** [comment]
re: sqlite that's not a sqlite issue, that's a cgo mangling issue w/ go_sqlite. Annoying.

**Comment by theAeon** [comment]
> Correction: mine contains that ifdef. I guess you can nest them because by calling something with extern "C" it auto-undefines it.
> 
> What I don't quite know is why it requires the wrapping outside, but it does fix the problem.

correction: you absolutely cannot do that, there are template directives above the ifdef. msvc is just being strange.

**Comment by Martchus** [comment]
I'd actually be curious whether it works with LLVM/Clang/libc++/Qt as provided by the MSYS2 mingw-w64 packaging. Their CLANGARM64 environment is quite similar to my build environment (except that everything is built on Windows and that some concrete versions/flags/patching differs).

**Comment by Martchus** [comment]
And feel free to create a PR with MSVC workarounds. If you make sure it still compiles with different compilers it would probably be mergable and then you wouldn't need to keep around your own branch.

**Comment by Martchus** [comment]
I achieved a working build that no longer hangs - at least in my WINE/qemu test setup. All I had to do was removing one flag when building Qt:

```
export CFLAGS=${CFLAGS//-D_FORTIFY_SOURCE=3/}
export CXXFLAGS=${CXXFLAGS//-D_FORTIFY_SOURCE=3/}
```

I didn't have to remove the flag when building my test app or when building mingw-w64 or and other dependencies. So I guess I'll workaround this by compiling Qt again with this flag removed. Note that setting the flag to 2 or 1 also doesn't work; it needs to be removed completely.

I also attached a development build of another (much simpler) app I used to reproduce/debug this (a simple Qt-based [password manager](https://github.com/Martchus/passwordmanager)): [passwordmanager.zip](https://github.com/user-attachments/files/22350935/passwordmanager.zip)

I would be curious whether it works natively on aarch64. I would hope so considering it works in my emulator setup now.

**Comment by Martchus** [comment]
I also created a build of Syncthing Tray: [syncthingtray.zip](https://github.com/user-attachments/files/22351703/syncthingtray.zip)

Unfortunately it doesn't work. Syncthing Tray now hangs much later, though. I have already removed `-D_FORTIFY_SOURCE=3` when building Syncthing Tray (and not just when building Qt 6 base). I probably have to remove that flag everywhere.

**Comment by Martchus** [comment]
After rebuilding c++utilities/qtutilities/qtforkawesome without fortify I was also able to make a working build of Syncthing Tray: [syncthingtray.zip](https://github.com/user-attachments/files/22370615/syncthingtray.zip)

Feel free to test it.

I don't think this is the end of the matter - even if this build works. We either have:

1. A regression in mingw-w64 13 (compared to 12) or LLVM/Clang 20.1.7/20.1.8 (compared to 19.1.7) that comes into play when `-D_FORTIFY_SOURCE=1` is used. There were [changes in mingw-w64 13](https://github.com/mingw-w64/mingw-w64/compare/v12.0.0...v13.0.0) that involve fortifying, e.g. in `stdio.h`. (In mingw-w64 code the `_FORTIFY_SOURCE` macro is "converted" to `__MINGW_FORTIFY…` macros so one has to search for relevant pieced of code by the latter.)
2. Undefined behavior in `qtbase` (and in at least one of `c++utilities`/`qtutilities`/`qtforkawesome` as I also had to rebuild those for Syncthing Tray to work again) that is exposed by updating mingw-w64 and LLVM/Clang/libc++ when fortifying is enabled. The first step to investigating this would probably be compiling individual source files with fortifying to narrow down the location of the problematic code.

By the way, I removed all broken builds from the release sections of my repos on GitHub. I'll upload aarch64-w64-mingw32 builds again on the next releases.

**Comment by Martchus** [comment]
I dared to provide an ARM build for v2.0.2 again. In my tests with WINE/qemu it hangs as soon as one uses the launcher but probably that's just due to my test setup. Otherwise it doesn't hang anymore in my test setup. 

By the way, it would be great if you could share the concrete commands you used to build this with MSVC. I assume some use of vcpkg was involved? This might be useful to others trying to build this on their own. (I also have some documentation for building with MSVC in the c++utilities repo but it doesn't cover aarch64 so far and it might not be the best/simplest approach. PRs for improving the build documentation in c++utilities are welcome.)

**Comment by theAeon** [comment]
I'm pretty sure it was just a clean cmake build and install-but there may have been a tiny tweak to cpputils, ill have to look.

I was using vcpkg for random non-qt dependencies, manually, but not as a toolchain file.

**Comment by Martchus** [comment]
Can someone provide any feedback on my ARM binaries after what I wrote in my previous comment?

**Comment by minzdrav** [comment]
Hi @Martchus 
Thank you. I can run it successfully! Look at screenshot.

<img width="1712" height="1508" alt="Image" src="https://github.com/user-attachments/assets/79c38a7f-6ea7-4b02-8dd6-02684dd631b2" />

**Comment by Martchus** [comment]
Great - and did you also use the launcher of Syncthing Tray to launch `syncthing.exe`? I'm asking because that didn't work in my WINE/qemu test setup.

**Comment by minzdrav** [comment]
Yes.

<img width="1933" height="399" alt="Image" src="https://github.com/user-attachments/assets/ebe35f4e-3474-4489-8b46-8318f7e59cc4" />

**Comment by theAeon** [comment]
> I'm pretty sure it was just a clean cmake build and install-but there may have been a tiny tweak to cpputils, ill have to look.
> 
> I was using vcpkg for random non-qt dependencies, manually, but not as a toolchain file.

Just PR'd, sorry.
